### PR TITLE
Cities Storage

### PR DESCRIPTION
**Pull request by PiotrChlapinskiBinarapps**
Now user if add new city to list and restart app, his selected weathers for cities are still in app.

### CODE REVIEW

File: Weather App/controller/ViewController.swift [hunk 1]
```diff
@@ -4,17 +4,23 @@ import UIKit
 class ViewController: UIViewController {
```
Comment (jarekdominiak) [inline Weather App/controller/ViewController.swift]: Could you name this ViewController? `WeathersListViewController` or something else

File: Weather App/viewModel/Storage.swift [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import Foundation
+
+public protocol Storage {
```
Comment (jarekdominiak) [inline Weather App/viewModel/Storage.swift]: Could you add description to protocol and functions?

File: Weather App/viewModel/WeatherForCityViewModel.swift [hunk 3]
```diff
@@ -4,25 +4,24 @@ import Keys
 
 protocol WeatherForCityViewModel {
```
Comment (jarekdominiak) [inline Weather App/viewModel/WeatherForCityViewModel.swift]: Could you add description to protocol, properties and functions?

File: Weather App/viewModel/WeatherForCityViewModel.swift [hunk 4]
```diff
@@ -4,25 +4,24 @@ import Keys
...
     var weathers: [WeatherForCity] { get }
+    var weatherForLocation: WeatherForCity? { get }
     func getWeatherForCity(cityName: String) async
     func getWeatherForCoordinate(coordinate: CLLocationCoordinate2D) async
 }
 
 class WeatherForCityViewModelImpl: WeatherForCityViewModel {
     var weathers: [WeatherForCity] = []
+    var weatherForLocation: WeatherForCity?
```
Comment (jarekdominiak) [inline Weather App/viewModel/WeatherForCityViewModel.swift]: CityWeather will be better, what do you think?

File: Weather App/viewModel/WeatherForCityViewModel.swift [hunk 5]
```diff
@@ -4,25 +4,24 @@ import Keys
...
 
     func getWeatherForCity(cityName: String) async {
-        let city = cityName.removeDiacratics()
-        let isContained = weathers.contains {
-            $0.name.removeDiacratics() == city
-        }
-        guard !isContained, let requestURL = Bundle.main.object(forInfoDictionaryKey: "WeatherRequestURL") as? String,
-              let url = URL(string: requestURL + "?q=\(city)" +
+        let cityName = cityName.removeDiacratics()
```
Comment (jarekdominiak) [inline Weather App/viewModel/WeatherForCityViewModel.swift]: Maybe there will be better to store cities names in storage without diacritics. We can also store two values inside a model, one for displaying and second for API. Btw, here is typo, should be `removeDiacritics `

File: Weather App/controller/ViewController.swift (hunk 1 above)
Comment (PiotrChlapinskiBinarapps) [reply Weather App/controller/ViewController.swift]: 👍

File: Weather App/viewModel/Storage.swift (hunk 2 above)
Comment (PiotrChlapinskiBinarapps) [reply Weather App/viewModel/Storage.swift]: 👍

File: Weather App/viewModel/WeatherForCityViewModel.swift (hunk 3 above)
Comment (PiotrChlapinskiBinarapps) [reply Weather App/viewModel/WeatherForCityViewModel.swift]: 👍

File: Weather App/viewModel/WeatherForCityViewModel.swift (hunk 4 above)
Comment (PiotrChlapinskiBinarapps) [reply Weather App/viewModel/WeatherForCityViewModel.swift]: 👍

File: Weather App/viewModel/WeatherForCityViewModel.swift (hunk 5 above)
Comment (PiotrChlapinskiBinarapps) [reply Weather App/viewModel/WeatherForCityViewModel.swift]: 👍

File: Weather App/WeathersList/viewModel/CitiesRepository.swift [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    public func getWeatherForCoordinate(_ coordinate: CLLocationCoordinate2D) async -> CityWeather? {
+        guard let url = URL(string: requestUrl + "?lat=\(coordinate.latitude)" +
+            "&lon=\(coordinate.longitude)&appid=\(Weather_AppKeys().openWeatherApiKey)&units=metric")
+        else {
+            return nil
+        }
+        return await getWeather(url: url, isFromLocation: true)
+    }
+
+    public func getWeather(url: URL, isFromLocation _: Bool) async -> CityWeather? {
```
Comment (jarekdominiak) [inline Weather App/WeathersList/viewModel/CitiesRepository.swift]: private?

File: Weather App/WeathersList/viewModel/CitiesRepository.swift [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    public func getWeatherForCity(_ city: City) async -> CityWeather? {
+        guard let url = URL(string: requestUrl + "?q=\(city.plainName)" +
+            "&appid=\(Weather_AppKeys().openWeatherApiKey)&units=metric")
+        else {
+            return nil
+        }
+        return await getWeather(url: url, isFromLocation: false)
+    }
+
+    public func getWeatherForCoordinate(_ coordinate: CLLocationCoordinate2D) async -> CityWeather? {
```
Comment (jarekdominiak) [inline Weather App/WeathersList/viewModel/CitiesRepository.swift:27]: do you need public access here? I think, internal will be enough

**Review by jarekdominiak — APPROVED** [review]

File: Weather App/WeathersList/viewModel/CitiesRepository.swift (hunk 7 above)
Comment (PiotrChlapinskiBinarapps) [reply Weather App/WeathersList/viewModel/CitiesRepository.swift:27]: 👍

**Pull request merged**